[IPPInAppFeedback] Orders screen banner placement

iamgabrielma opened the pull request:
Closes: #8636 
Part of: https://github.com/woocommerce/woocommerce-ios/issues/8530

## Description

This PR adds the in-app feedback banner to the Order view, as well as the buttons to launch a survey and dismiss the banner.

At the moment the banner is not dismissible. Dismiss will be part of a [different PR](https://github.com/woocommerce/woocommerce-ios/issues/8589).

## Changes
- Adds a `.IPP` case as new `FeedbackType`
- Upon configuring `OrderListViewController`, we set the existing top banner view (generally used to display the Order Creation banner, or an error) to a new IPP banner survey. This uses the existing `TopBannerView` and `TopBannerViewModel` components.
- We have a new Publisher `hideIPPFeedbackBanner`, so we can listen to changes and set priorities on the different banners that should display in the view.
- The banner has an `ActionButton` that will launch the feedback survey via the existing `SurveyCoordinatingController` component.

## Testing instructions

1. Go to Menu > Orders > See the banner at the top of the screen
2. Tap "Share Feedback", a WebView showing a survey form will shop up. Scroll down and tap "Finish Survey", a "Feedback Sent" view will be shown and you'll be redirected to the Orders page.

## Screenshots
![Simulator Screen Recording - iPhone 11 Pro - 2023-01-16 at 12 40 23](https://user-images.githubusercontent.com/3812076/212600007-00e6d0de-bff9-4cf7-8b0e-2a52aa45ed1d.gif)


---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

iamgabrielma committed:
Add new IPPFeedback topBannerView to OrderList view

iamgabrielma committed:
Create and configure IPP feedback topBannerView

iamgabrielma committed:
Make the top banner as per design specs

iamgabrielma committed:
Add NSLocalizedStrings

iamgabrielma committed:
Display feedback survey when tapped

iamgabrielma committed:
Add IPPFeedback as new FeedbackType

iamgabrielma committed:
Add feature flag case to Unit Tests

iamgabrielma committed:
Dispatch loadFeedbackVisibility action

toupper commented on Storage/Storage/Model/FeedbackType.swift:
```diff
@@ -16,4 +16,8 @@ public enum FeedbackType: String, Codable {
+
+    /// Identifier for the In-Person Payments feedback survey
+    ///
+    case IPPFeedback
```
Minor thing here, if the enum type name already contains "Feedback", I think we can omit it here: `case IPP`, as in the case of the `ordersCreation`

iamgabrielma replied:
Changed on https://github.com/woocommerce/woocommerce-ios/pull/8640/commits/83835c4f1a1a93249f564d4d2bd0217ccf3547b4

toupper commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift:
```diff
@@ -143,6 +148,12 @@ final class OrderListViewModel {
+
+        if ServiceLocator.featureFlagService.isFeatureFlagEnabled(.IPPInAppFeedbackBanner) {
+            // TODO: Remove. Temporarily set to false so we can debug the IPPFeedback banner easily.
+            hideOrdersBanners = false
```
I find a bit dangerous to keep this in the PR, even if we add the TODO comment. If it goes to `trunk` it would disable the banner, which is not what we want

iamgabrielma replied:
You're right, that makes sense. I'll experiment with the `@Published` properties to find a better way to display/hide these. 

I see we're using `Publishers.CombineLatest(:)` so the view model internal state figures out which banner should assign to the view, we could pass the `$hideIPPFeedbackBanner` binding as well there and use `CombineLatest3(:)` to figure this one out.

toupper commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift:
```diff
@@ -165,10 +176,12 @@ final class OrderListViewModel {
-        loadOrdersBannerVisibility()
 
         if ServiceLocator.featureFlagService.isFeatureFlagEnabled(.IPPInAppFeedbackBanner) {
+            loadIPPFeedbackBannerVisibility()
```
I have a question here, don't we load the orders banner if the IPP Feedback is enabled?

iamgabrielma replied:
Good question, after thinking further about it I was unsure how to fully deal with the Order Creation banner if IPP is enabled. 

As per discussion at p1673883711765819?thread_ts=1673876649.359489&cid=C025A8VV728-slack-C025A8VV728 we're opting for:

- Prioritise showing the IPP survey. Show the Order Creation banner only when the IPP survey is not displayed.
- If IPP survey is dismissed, ideally we shouldn't show Order Creation immediately, unless that's the default behavior and changing it involves a certain level of complexity (as we're targeting to release this survey as soon as possible, then iterate later if needed).

toupper commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift:
```diff
@@ -185,6 +198,15 @@ final class OrderListViewModel {
     }
 
+    func dismissIPPFeedbackBanner() {
+        print("dismiss tapped")
```
Can we delete these prints here, as this is likely going to `trunk`?

iamgabrielma replied:
Definitely, removed on https://github.com/woocommerce/woocommerce-ios/pull/8640/commits/ff7a3eeb711ac49f00ce9b55f0d78cc84e277a26

toupper commented:
Great job @iamgabrielma! I added a few comments regarding the code.
Apart from that, I have a question about the design. It seems like the banner message -"_Rate your In-Person Payments experience_"- does not go under the x button in the [original design](T16Ov1hzHWLJSOr8iN1BgR-fi-202%3A8998&t=Jr8EhOCOFMc4SeQm-0), unlike in your implementation:

<img src="https://user-images.githubusercontent.com/1864060/212654012-35d0b14f-26a1-40a7-9621-94b3d718231d.png" width="375">

Is that something we have to fix?

iamgabrielma committed:
Rename IPP FeedbackType case

iamgabrielma committed:
Remove print statements

iamgabrielma committed:
Add IPP vs Order Creation banner priority

iamgabrielma committed:
Remove dismiss banner logic

The IPP banner dismiss logic is not part of this PR. Removing for simplicity and to reduce scope.

iamgabrielma committed:
Make hideOrdersBanners always true on init

iamgabrielma committed:
Unit Tests for IPP banner visibility

iamgabrielma committed:
Resolve IPP vs Order Create banner unit tests conflicts

iamgabrielma committed:
Change initial IPP banner visibility state

iamgabrielma committed:
Update banner status on each run (for testing)

iamgabrielma commented:
Thanks for your review @toupper , I addressed the comments and this is ready for review again 👍 

> I have a question about the design. It seems like the banner message -"Rate your In-Person Payments experience"- does not go under the x button in the [original design](https://github.com/woocommerce/woocommerce-ios/pull/T16Ov1hzHWLJSOr8iN1BgR-fi-202%3A8998&t=Jr8EhOCOFMc4SeQm-0), unlike in your implementation. Is that something we have to fix?

I missed that one! This UI tweak and others is something I'd leave for a different PR ( [tracked here](https://github.com/woocommerce/woocommerce-ios/issues/8674) ), as we're using an existing component I wanted to work through its functionality first 👯

toupper commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift:
```diff
@@ -365,18 +400,18 @@ extension OrderListViewModel {
+            .map { hasError, hasDismissedIPPFeedbackBanner, hasDismissedOrdersBanners -> TopBanner in
 
-                if hasError {
+                guard !hasError else {
```
So beautiful code 😉

toupper commented:
Great job @iamgabrielma! Impressive how you are progressively tackling more complex features 👏 :shipit:
